src/climate_indices/compute.py

```diff
@@ -743,3 +743,267 @@ def transform_fitted_gamma(
+        raise ValueError(message)
+
+    else:
+
+        time_steps_per_year = values.shape[1]
+        if (time_steps_per_year != 12) and (time_steps_per_year != 366):
+            message = "Invalid shape of input data array: {shape}".format(shape=values.shape)
+            _logger.error(message)
+            raise ValueError(message)
+
+    # determine the end year of the values array
+    data_end_year = data_start_year + values.shape[0]
+
+    # make sure that we have data within the full calibration period,
+    # otherwise use the full period of record
+    if (calibration_start_year < data_start_year) or \
+            (calibration_end_year > data_end_year):
+        calibration_start_year = data_start_year
+        calibration_end_year = data_end_year
+
+    # get the year axis indices corresponding to
+    # the calibration start and end years
+    calibration_begin_index = calibration_start_year - data_start_year
+    calibration_end_index = (calibration_end_year - data_start_year) + 1
+
+    # get the values for the current calendar time step
+    # that fall within the calibration years period
+    calibration_values = values[calibration_begin_index:calibration_end_index, :]
+
+    # the values we'll compute and return
+    probabilities_of_zero = np.zeros((time_steps_per_year,))
+    c = np.zeros((time_steps_per_year,))
+    scales = np.zeros((time_steps_per_year,))
+    locs = np.zeros((time_steps_per_year,))
+
+    # compute the probability of zero and Pearson
+    # parameters for each calendar time step
+    # TODO vectorize the below loop? create a @numba.vectorize() ufunc
+    #  for application over the second axis
+    for time_step_index in range(time_steps_per_year):
+
+        # get the values for the current calendar time step
+        time_step_values = calibration_values[:, time_step_index]
+
+        # count the number of zeros and valid (non-missing/non-NaN) values
+        number_of_zeros, number_of_non_missing = \
+            utils.count_zeros_and_non_missings(time_step_values)
+
+        # make sure we have at least four values that are both non-missing (i.e. non-NaN)
+        # and non-zero, otherwise use the entire period of record
+        if (number_of_non_missing - number_of_zeros) < 4:
+
+            # we can't proceed, bail out using zeros
+            continue
+
+        # calculate the probability of zero for the calendar time step
+        probability_of_zero = 0.0
+        if number_of_zeros > 0:
+
+            probability_of_zero = number_of_zeros / number_of_non_missing
+
+        # get the estimated parameters, if we have
+        # more than three non-missing/non-zero values
+        if (number_of_non_missing - number_of_zeros) > 3:
+
+            # get the fisk parameters for this time
+            # step's values within the calibration period
+            c[time_step_index], scales[time_step_index], locs[time_step_index], = \
+                scipy.stats.fisk.fit(time_step_values)
```

Comment:
Just had a thought about this last night. How good is this fit that uses Maximum Likelihood Estimation? For pearson3, the L-moments method was used (which needed to be implemented because scipy stats doesn't have it as an option), and for gamma, the Maximum Likelihood Estimation was used, but via the Thom method, which also isn't available through scipy.stats. I have no experience in judging various parameter estimations, but we probably should have some sort of justification for which approach is used. I think having an existing, well-cited paper being a good standard.

Reply:
I originally wanted to add this since the Spanish authors who are most famous for SPEI used this distribution for much of their work. They have provided [an implementation in R](https://cran.r-project.org/web/packages/SPEI/SPEI.pdf). The methodology is nicely described [here](https://www.researchgate.net/publication/291185005_SPEI-Based_Spatiotemporal_Analysis_of_Drought_in_Haihe_River_Basin_from_1961_to_2010) (page 3)

Here is an interesting [goodness of fit analysis](https://rstudio-pubs-static.s3.amazonaws.com/202764_0f164cddb7fd415eae1389ea27cec447.html)

Reply:
I love papers like this! Very clearly lays out the algorithm, both in words and in equations, and also provides names for some of the steps they use. It is clear from this paper that an L-moments method is used (but seems different from the L-moments used for pearson III). The formulation is also different enough to give me pause, but that just might be that the scipy documentation is showing a standardized form while the paper isn't. It is going to take a bit of algebra to cross-check the paper's L-moments with the parameters used in scipy's implementation.

On a science note, I am a bit confused by the "non-negative" aspect of this distribution, given that Pt - PET can be negative, but I guess that is what the "loc" (in scipy) or "gamma" (in paper) can account for?